semantic-analysis/src/Analysis/Syntax.hs

```diff
@@ -22,105 +20,79 @@ module Analysis.Syntax
 , parseNode
 ) where
 
+import qualified Analysis.Carrier.Statement.State as S
 import           Analysis.Effect.Domain
-import           Analysis.Effect.Env (Env, bind)
+import           Analysis.Effect.Env (Env, bind, lookupEnv)
 import           Analysis.Effect.Store
-import           Analysis.Name (Name, formatName, nameI)
-import           Control.Applicative (Alternative(..), liftA3)
+import           Analysis.File
+import           Analysis.Module
+import           Analysis.Name (Name, name, nameI)
+import           Analysis.Reference as Ref
+import           Control.Applicative (Alternative (..), liftA3)
 import           Control.Effect.Labelled
+import           Control.Effect.Throw (Throw, throwError)
 import           Control.Monad (guard)
+import           Control.Monad.IO.Class
 import qualified Data.Aeson as A
 import qualified Data.Aeson.Internal as A
 import qualified Data.Aeson.Parser as A
 import qualified Data.Aeson.Types as A
 import qualified Data.ByteString.Lazy as B
 import           Data.Function (fix)
 import qualified Data.IntMap as IntMap
-import           Data.Monoid (First(..))
-import           Data.Text (Text, pack, unpack)
+import           Data.List (sortOn)
+import           Data.List.NonEmpty (NonEmpty, fromList, toList)
+import           Data.Monoid (First (..))
+import qualified Data.Set as Set
+import           Data.String (IsString (..))
+import           Data.Text (Text, pack)
+import qualified Data.Text as Text
 import qualified Data.Vector as V
+import qualified System.Path as Path
 
-class Syntax rep where
-  iff :: rep -> rep -> rep -> rep
-  noop :: rep
-
-  bool :: Bool -> rep
-  string :: Text -> rep
-
-  throw :: rep -> rep
-
-  let_ :: Name -> rep -> (rep -> rep) -> rep
-
-
--- Pretty-printing
-
-newtype Print = Print { print_ :: ShowS }
-
-instance Show Print where
-  showsPrec _ = print_
-
-instance Semigroup Print where
-  Print a <> Print b = Print (a . b)
-
-instance Monoid Print where
-  mempty = Print id
-
-instance Syntax Print where
-  iff c t e = parens (str "iff" <+> c <+> str "then" <+> t <+> str "else" <+> e)
-  noop = parens (str "noop")
-
-  bool b = parens (str (if b then "true" else "false"))
-  string = parens . text
-
-  throw e = parens (str "throw" <+> e)
-
-  let_ n v b = parens (str "let" <+> name n <+> char '=' <+> v <+> str "in" <+> b (name n))
-
-str :: String -> Print
-str = Print . showString
-
-text :: Text -> Print
-text = str . unpack
-
-char :: Char -> Print
-char = Print . showChar
-
-parens :: Print -> Print
-parens p = char '(' <> p <> char ')'
-
-(<+>) :: Print -> Print -> Print
-l <+> r = l <> char ' ' <> r
-
-infixr 6 <+>
-
-name :: Name -> Print
-name = text . formatName
+data Term
+  = Var Name
+  | Noop
+  | Iff Term Term Term
+  | Bool Bool
+  | String Text
+  | Throw Term
+  | Let Name Term Term
+  | Import (NonEmpty Text)
+  deriving (Eq, Ord, Show)
```

Comment:
The tagless final class, and the `Print` and `Interpret` instantiations of it are gone, because defining an `Ord` instance for `Interpret` would have been agonizing. Instead, we parse into a first-order, initial encoding: plain old data.

Reply:
Plain old data is (usually) my favorite kind of data!

Reply:
It's just, it was so nice! But too clever by half.